### ISSUE TITLE
Store images somewhere else than S3

### ISSUE THREAD

**Issue by scastiel** [spliit-app/spliit#65]
Since #63, users can attach images to expenses, but they can be stored only on S3. If someone has the need to store images somewhere else (locally, on an FTP server…), feel free to comment here with your need and we’ll see how to implement it.

**Comment by mertd** [comment]
A relatively simple change could be to support S3 providers other than AWS and allow configuring the endpoint via an additional environment variable like `S_UPLOAD_ENDPOINT`.

Support for additional storage methods may not be necessary beyond this. There are FOSS options for S3 compatible storage so there is no vendor lock-in. Focusing storage options on a single method may help keeping complexity lower, too.

**Comment by justcallmelarry** [comment]
Looks like next-s3-upload has support for the endpoint already:
https://next-s3-upload.codingvalue.com/bucket-config#bucket-endpoint

**Comment by ChristopherJohnston** [comment]
support for storage on a local file system might be useful for homelab deployments

**Comment by justcallmelarry** [comment]
> support for storage on a local file system might be useful for homelab deployments

it could, but you should also just be able to use for example [Minio](https://min.io/) in order to reach the same functionality, and since you are running multiple containers already (postgres+app) i don't think the overhead of adding for example Minio would hurt too much.
Then you can choose where to store the Minio volumes if you wish

**Comment by justcallmelarry** [comment]
I tried getting Minio running locally and got it to work with uploading the image to the bucket, but not with then showing it.

I think you would have to make it work with serving images via presigned URLs (in addition to uploading via presigned URLs) in order to get it to work. I can continue looking into it another day if this would be a suitable solution?

**Comment by mertd** [comment]
It seems we both gave it a try in parallel. :)

I'd be glad to hear your feedback to my attempt: #71

**Comment by justcallmelarry** [comment]
> It seems we both gave it a try in parallel. 🙂
> 
> I'd be glad to hear your feedback to my attempt: #71

Looks very similar to my try! Except yours handles the next config parts way better than my solution.
It seems my issue was that I did not configure my Minio bucket correctly, so the only thing I would maybe say could be added is a hint in the documentation about Minio and how to setup the bucket. But I'm also unsure if that actually should be in the documentation for Spliit… so I did not leave it as a comment on the PR.

Very nice work!

**Comment by ChristopherJohnston** [comment]
> I tried getting Minio running locally and got it to work with uploading the image to the bucket, but not with then showing it.
> 
> 
> 
> I think you would have to make it work with serving images via presigned URLs (in addition to uploading via presigned URLs) in order to get it to work. I can continue looking into it another day if this would be a suitable solution?

Regarding exposing to the internet, one of my concerns is having too many ports open on a home server so I have been trying to embed spliit as an add-on for homeassistant so that I can use a single open port as the entry point or their nabucasa proxy - it would then be available from the homeassistant app and this along with all of my other tools are available in in one place.

This would solve the public URL problem too

**Comment by justcallmelarry** [comment]
> > I tried getting Minio running locally and got it to work with uploading the image to the bucket, but not with then showing it.
> > I think you would have to make it work with serving images via presigned URLs (in addition to uploading via presigned URLs) in order to get it to work. I can continue looking into it another day if this would be a suitable solution?
> 
> Regarding exposing to the internet, one of my concerns is having too many ports open on a home server so I have been trying to embed spliit as an add-on for homeassistant so that I can use a single open port as the entry point or their nabucasa proxy (see #56). The app would then be available from the homeassistant app and this along with all of my other tools are available in one place.
> 
> This would solve the public URL problem too as the MinIO url would be considered local.

Yeah, I run everything behind a reverse proxy, so I have only port 443 open to the internet, but using different sub-domains. But I do see the use case of being able to just source them from a static file directory.

**Comment by webzit** [comment]
I would also prefer local storage for home-lab environments. MinIO makes it more complex overall and another service that has to be hardened against external threats.

**Comment by LIRIKKER** [comment]
Should look to keep documents stored locally as for home-lab users looking to keep everything in-house would make the most sense.

Once the docker hub image is available, you could look at integrating with other applications.
As most home-lab users are using paperless-ngx, could look to use the integration for this with Spliit?

**Comment by swiftbird07** [comment]
@justcallmelarry could you please share how you configured Minio so the images are actually displayed in Spliit (PUT works fine).

**Comment by swiftbird07** [comment]
Any updates on this?

**Comment by DanielAtkinson** [comment]
@swiftbird07 I had the problem of intergrating the minio docker image with the spliit docker image also. I could upload an image with PUT, but when the split application tried to do a GET it failed.

For me, one of the solutions was to open the port 9000 ( The port minio was using for the buckets ) on my docker machine using UFW. There could have been other configuration changes that contributed to the application working I can share if it would help? I modified the composer file to use the minio image so they share a network.

If you are still interested in fixing this, I'm happy to share what I have done! :) It took me a while to trouble shoot this.

**Comment by swiftbird07** [comment]
> If you are still interested in fixing this, I'm happy to share what I have done! :) It took me a while to trouble shoot this.

Yes I would appreciate that. Especially how did you set up the permissions etc?

**Comment by LnLcFlx** [comment]
Thanks for this minio workaround, however I have trouble getting it to work.
I followed your steps but I get a "Error while uploading document. Something wrong happened..." in the UI.
Both in MinIO and Spliit logs I see no error.

I created a bucket called `spliit` and an access key with the name `spliit`.
The access key is `ABC` and the secret `DEFGHIJK`.
I also changed expiry to below 1y and bucket access policy to public.
The MinIO region is set to `us-east-1`.
Both containers are in the same docker network and the `container_name` of MinIO is set to `spliit-s3`.
I then added the env vars to spliit
```
NEXT_PUBLIC_ENABLE_EXPENSE_DOCUMENTS=true
S3_UPLOAD_KEY=ABC
S3_UPLOAD_SECRET=DEFGHIJK
S3_UPLOAD_BUCKET=spliit
S3_UPLOAD_REGION=us-east-1
S3_UPLOAD_ENDPOINT=http://spliit-s3:9000
```
This should work right?

Also in your comment above you wrote
```
S3_UPLOAD_KEY=Name-of-access-key-generated-in-minio
S3_UPLOAD_SECRET=Name-of-access-key-generated-in-minio
```
two times `Name-of-...`

**Comment by 6543** [comment]
well the aplication does not proxy but let the client push directly to the s3 bucked

@scastiel how can i contact you privately ? matrix ? or mail ?

**Comment by suuuehgi** [comment]
> it could, but you should also just be able to use for example [Minio](https://min.io/) in order to reach the same functionality, and since you are running multiple containers already (postgres+app) i don't think the overhead of adding for example Minio would hurt too much.
Then you can choose where to store the Minio volumes if you wish

I think it's downright cheeky to recommend MinIO when Spliit could just write the data directly to disk. MinIO ties up at least 300 MB of RAM when idle, and as an image, it's just as big as Spliit itself.

**Comment by BartschLabs** [comment]
@dDiverVS & @LnLcFlx, I'm trying to get working with RustFS what you got working with Minio.  

I'm pretty close.  I had the same error one of you had with the http S3 endpoint, so I used an https S3 endpoint and I can upload, but Spliit can't read the uploaded docs.  I'm wondering if you've discovered any more while using Minio to help me out.

My vars:
```
NEXT_PUBLIC_ENABLE_EXPENSE_DOCUMENTS=true
S3_UPLOAD_KEY=****
S3_UPLOAD_SECRET=****
S3_UPLOAD_BUCKET=spliit
S3_UPLOAD_REGION=us-east-1
S3_UPLOAD_ENDPOINT=https://s3.mydomain.com
```

My bucket has all of the images I attached.  They're presented as broken links to my Spliit expenses.  I get this error in the Spliit container log every time Spliit tries to access a file in the bucket:

`upstream image https://s3.mydomain.com/spliit/document-2026-08-27T22:28:27.766Z-0VPy7HYg0Qja8BOjIm19F.jpg hostname resolved to private IP ["192.168.1.3"] If this is expected and you understand SSRF risk, use images.dangerouslyAllowLocalIP = true to continue.`

While it's not advisable for a production environment, I tried to add NEXT_PUBLIC_IMAGES_DANGEROUSLY_ALLOW_LOCAL_IP=true, but that env var doesn't seem to be recognized by the container image.

Ideas?

**Comment by dDiverVS** [comment]
Hi @BartschLabs , sorry for the late response, I think you pretty much run into the same issue as I had with the origin of the requests being a private instead of a public IP.
I see from your log that your domain is being resolved as a private IP "192.168.1.3", is it possible at all to use your WAN IP instead? Otherwise I think you will need to check the configuration files and manually update the images.dangerouslyAllowLocalIP value to true.

Also it might be worth a try to use the variable IMAGES_DANGEROUSLY_ALLOW_LOCAL_IP instead as the NEXT_PUBLIC_ part is deprecated according to the documentation.

**Comment by BartschLabs** [comment]
> is it possible at all to use your WAN IP instead? 

Hi @dDiverVS, thanks for the suggestion.  It is not easy to use my WAN IP.  I use a Caddy reverse proxy that straddles my LAN 192.168.4.0 and my DMZ 192.168.1.0.  External DNS is resolving my S3 domain name to my WAN IP.  Internal DNS is resolving my S3 domain name to the Caddy interface on my DMZ.  It works fine, except for this.

> Otherwise I think you will need to check the configuration files and manually update the images.dangerouslyAllowLocalIP value to true.

I don't know where the configuration files are.  I'm only using docker environment variables.

> Also it might be worth a try to use the variable IMAGES_DANGEROUSLY_ALLOW_LOCAL_IP instead as the NEXT_PUBLIC_ part is deprecated according to the documentation.

I tried, but it has no effect.  It's unclear why I have to use NEXT_PUBLIC_ENABLE_EXPENSE_DOCUMENTS=true instead of ENABLE_EXPENSE_DOCUMENTS=true if NEXT_PUBLIC is deprecated.

**Comment by dDiverVS** [comment]
@BartschLabs I think then you can test to add the value dangerouslyAllowLocalIP: true, to [this file](https://github.com/spliit-app/spliit/blob/ae35cd6d758a287ca92c5529fcf6f6dcf457dc3d/next.config.mjs#L31) and compile it again. If you are using the precompiled docker image I don't know if you can add it somehow.

The section in line 31 should look afterwards like this:
```
images: {
    remotePatterns,
    dangerouslyAllowLocalIP: true
}
```


Regarding the NEXT_PUBLIC_ variables both should work but it is recommended to use the new ones, you can see it here: https://github.com/spliit-app/spliit#configuration

> The old variables still work — the runtime variant simply takes precedence when both are set, so there is nothing you have to change immediately. They remain the right choice if you build your own image and want a setting baked in. To migrate, drop the NEXT_PUBLIC_ prefix and set the variable wherever your container gets its environment.

Hope this helps 😄

**Comment by BartschLabs** [comment]
@dDiverVS, thanks again for the help.

I prefer to use the pre-built image, but I might make an exception.  However, for some reason I no longer have the error, though I did put the parameter in the runtime copy of next.config.js while troubleshooting, but figured the file would be lost during any of the redeployment I did of the image.  I'll have to check if my change to the file persisted.

I still have broken links displayed for the files I've uploaded.  The links appear to to be correct, but when I try one I get an error from next that "url" parameter is not allowed.  Maybe RustFS is not accepting something in the url parameter of the string.  I see that there are dimensions appended to the parameter.  These might be causing RustFS to error.
